File: julia/MOLE.jl/examples/elliptic1D.jl
"""
A Julia script that solves the 1D Poisson's equation with Robin boundary conditions using mimetic operators.
"""

using Plots
using MOLE

# Domain limits
west = 0.0
east = 1.0

k = 6 # operator order of accuracy
m = 2*k + 1 # minimum number of cells to attain desired accuracy
dx = (east-west)/m # step length

L = lap(k,m,dx)

# Impose Robin boundary condition on laplacian operator
a = 1.0
b = 1.0
L = L + robinBC(k,m,dx,a,b)

# 1D staggered grid
grid = [west; (west+(dx/2)):dx:(east-(dx/2)); east]

# RHS
U = exp.(grid)
U[1] = 0
U[end] = 2*exp(1)

U = L\U

# Plot result
p = scatter(grid, U, label="Approximated")
plot!(p, grid, exp.(grid), label="Analytical")
plot!(p, xlabel="x", ylabel="u(x)", title="Poisson's equation with Robin BC")
display(p)
println("Press Enter to close the plot and exit.")
readline()

File: examples/matlab_octave/elliptic1D.m
% Solves the 1D Poisson's equation with Robin boundary conditions

clc
close all

addpath('../../src/matlab_octave')

west = 0;  % Domain's limits
east = 1;

k = 6;  % Operator's order of accuracy
m = 2*k+1;  % Minimum number of cells to attain the desired accuracy
dx = (east-west)/m;  % Step length

% Impose Robin BC on laplacian operator
a = 1;
b = 1;
dc = [a; a];
nc = [b; b];
v = [0; 0];
g = makeGrid('m', m, 'dx', dx, 'bc', struct('dc', dc, 'nc', nc));
L = lap(g, k);  % 1D Mimetic laplacian operator
[L_bc, ~] = addScalarBC(sparse(size(L,1), size(L,2)), zeros(size(L,1),1), k, g, v);
L = L + L_bc;

% 1D Staggered grid
grid = [west west+dx/2 : dx : east-dx/2 east];

% RHS
U = exp(grid)';
U(1) = 0;  % West BC
U(end) = 2*exp(1);  % East BC

U = L\U;  % Solve a linear system of equations

% Plot result
plot(grid, U, 'o')
hold on
plot(grid, exp(grid))
legend('Approximated', 'Analytical', 'Location', 'NorthWest')
title('Poisson''s equation with Robin BC')
xlabel('x')
ylabel('u(x)')
